### PR TITLE
Position modify refactor

### PR DESCRIPTION
**Pull request by bachstatter**
Create a new position-modified.ts file and function to handle position modifications in a seperate place.
Create discrete functions that handles, `postionOpened`,`positionClosed`, `positionModified`. 
Move calculations to calculations.ts file. 

Bugs fixed:
- Funding, we forgot to update funding index on the position after adding accrued funding
- Make sure totalVolume is updated on Trader and FuturesPosition
- Pnl for trader was incorrect in some cases
- Pnl for position closed was overriding previous pnls.
- We were only deducting fees when position was opened, not when modifying/ closing. I've now changed it to not include fees. The position entity do have fees frontend can easily calculate pnl by subracting fees and adding funding. In the future we could look into creating, realised vs unrealised pnl.
- In some cases we were adding funding to `feesPaidToSynthetix`. Funding is never paid to stakers.

### CODE REVIEW

**Comment by bachstatter** [comment]
I think I wanna dig a bit more tomorrow, especially related to avgEntryPrice. But can do that in a separate PR

**Comment by bachstatter** [comment]
I also want to add some more tests for multiple actions and track PNL along the way.. 
If we calculate like this:
```
event.params.lastPrice
      .minus(futuresPosition.avgEntryPrice)
      .times(futuresPosition.size)
```
Does it make sense to `futuresPosition.pnl.plus(newPnl);` or  should it just be `futuresPosition.pnl = newPnl`; since we're using the average price. Maybe my change to position closed doesn't make sense, and I should change modification instead..

File: v2/perps-v2/perps-subgraph/src/position-modified.ts [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+    event.params.size,
+    event.params.lastPrice,
+    event.params.margin
+  );
+  futuresPosition.netFunding = BigInt.fromI32(0);
+  futuresPosition.txHash = event.transaction.hash.toHex();
+  futuresPosition.totalVolume = calculateVolume(event.params.tradeSize, event.params.lastPrice);
+  if (
+    network === 'optimism-goerli' &&
+    event.block.number.gt(SIP2004_AND_2005_GOERLI_BLOCK_NUMBER)
```
Comment (fritzschoff) [inline v2/perps-v2/perps-subgraph/src/position-modified.ts:99]: `gte` no? But not super important I guess

File: v2/perps-v2/perps-subgraph/src/position-modified.ts [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+function handlePositionClosed(
+  event: PositionModifiedNewEvent,
+  futuresPosition: FuturesPosition,
+  trader: Trader,
+  synthetix: Synthetix
+): void {
+  const newPnl = calculatePnl(
+    event.params.lastPrice,
+    futuresPosition.avgEntryPrice,
+    futuresPosition.size // Note that it's important to use the size before we update it from the event. The updated size will be 0 since the position is closed
```
Comment (fritzschoff) [inline v2/perps-v2/perps-subgraph/src/position-modified.ts:135]: 👍

File: v2/perps-v2/perps-subgraph/src/position-modified.ts (hunk 1 above)
Comment (bachstatter) [reply v2/perps-v2/perps-subgraph/src/position-modified.ts:99]: I dont think I touched that logic. But yea happy to change it!

**Review by jmzwar — COMMENTED** [review]
Still confused why Pnl doesn't consider funding.

File: v2/perps-v2/perps-subgraph/src/calculations.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+export function calculateVolume(tradeSize: BigInt, lastPrice: BigInt): BigDecimal {
+  return tradeSize.times(lastPrice).div(BigInt.fromI32(10).pow(18)).abs().toBigDecimal();
+}
+
+export function calculateLeverage(size: BigInt, lastPrice: BigInt, margin: BigInt): BigInt {
+  return size.times(lastPrice).div(margin).abs();
+}
+
+export function calculatePnl(lastPrice: BigInt, avgEntryPrice: BigInt, size: BigInt): BigInt {
+  return lastPrice.minus(avgEntryPrice).times(size).div(BigInt.fromI32(10).pow(18));
```
[the remark this answers is not in the document]
Comment (bachstatter) [reply v2/perps-v2/perps-subgraph/src/calculations.ts:13]: Yea for now, pnl does not take funding or fees into account.
So to get the gross pnl on the frontend we need to:
`pnl - feesPaidToSynthetix + netFundind`

**Review by jmzwar — APPROVED** [review]

**Pull request merged**